sklearn/model_selection/_search.py

```diff
@@ -923,11 +923,10 @@ def evaluate_candidates(candidate_params, cv=None, more_results=None):
             self.best_params_ = results["params"][self.best_index_]
 
         if self.refit:
-            # we clone again after setting params in case some
-            # of the params are estimators as well.
```

Comment:
Two points: (1) I think we should keep a comment here (and at the other diff) explaining why the parameters are being cloned as well as the estimator. It feels like an unusual enough thing that it is worth explaining to people from the future. (2) doesn't the comment say that what is happening here should prevent the reported bug from happening? Did that just not work? Maybe someone who understands this better can help me.

Reply:
1- done
2- yes, but the code doesn't actually do what the comment guarantees, and we didn't have a test for it. The issue is basically that `set_params` doesn't `clone` before setting. So it first sets the parameters which are estimators, then it sets the parameters of those set estimators, which happen to be the same object instance as the one in the `param_grid`, hence changing the original instance.

Reply:
Aha! Thanks for explaining